plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteVectorDistanceFunction.java

```diff
@@ -0,0 +1,159 @@
+import static java.lang.Double.longBitsToDouble;
+import static java.util.Objects.requireNonNull;
+
+public final class RewriteVectorDistanceFunction
+        implements ProjectFunctionRule<JdbcExpression, ParameterizedExpression>
+{
+    private static final Capture<ConnectorExpression> LEFT_ARGUMENT = newCapture();
+    private static final Capture<ConnectorExpression> RIGHT_ARGUMENT = newCapture();
+
+    private final Pattern<Call> pattern;
+    private final String operator;
+
+    public RewriteVectorDistanceFunction(String functionName, String operator)
+    {
+        pattern = call()
+                .with(functionName().equalTo(new FunctionName(requireNonNull(functionName, "functionName is null"))))
+                .with(type().matching(type -> type == DOUBLE))
+                .with(argumentCount().equalTo(2))
+                .with(argument(0).matching(expression().capturedAs(LEFT_ARGUMENT).with(type().matching(RewriteVectorDistanceFunction::isArrayTypeWithRealOrDouble))))
+                .with(argument(1).matching(expression().capturedAs(RIGHT_ARGUMENT).with(type().matching(RewriteVectorDistanceFunction::isArrayTypeWithRealOrDouble))));
+        this.operator = requireNonNull(operator, "operator is null");
+    }
+
+    @Override
+    public Pattern<? extends ConnectorExpression> getPattern()
+    {
+        return pattern;
+    }
+
+    @Override
+    public Optional<JdbcExpression> rewrite(ConnectorExpression projectionExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Optional<ParameterizedExpression> leftExpression = rewrite(captures.get(LEFT_ARGUMENT), context);
+        if (leftExpression.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<ParameterizedExpression> rightExpression = rewrite(captures.get(RIGHT_ARGUMENT), context);
+        if (rightExpression.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new JdbcExpression(
+                "%s %s %s".formatted(leftExpression.get().expression(), operator, rightExpression.get().expression()),
+                ImmutableList.<QueryParameter>builder()
+                        .addAll(leftExpression.get().parameters())
+                        .addAll(rightExpression.get().parameters())
+                        .build(),
+                new JdbcTypeHandle(
+                        Types.DOUBLE,
+                        Optional.of("double"),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty())));
+    }
+
+    private static Optional<ParameterizedExpression> rewrite(ConnectorExpression expression, RewriteContext<ParameterizedExpression> context)
+    {
+        if (expression instanceof Constant constant) {
+            Object value = constant.getValue();
+            LongArrayBlock block = (LongArrayBlock) value;
```

Comment:
This allows to expose the block <-> type relation we have today - Is it possible for us to generate them via `Type` API ?